File: learning/cpp/02_control_flow.cpp
#include <iostream>
#include <vector>

// C++入門: 02 - 制御フロー

// プログラムの流れをコントロールするための構文を学びます。
// 条件分岐 (if, else if, else) や繰り返し (for, while) は、あらゆるプログラムの基本です。

int main() {
    // 1. if, else if, else文 (条件分岐)
    // -------------------------------------
    int temperature = 25;

    std::cout << "--- if文の例 --- " << std::endl;
    if (temperature >= 30) {
        std::cout << "暑いですね！" << std::endl;
    } else if (temperature >= 20) {
        std::cout << "快適な気温です。" << std::endl;
    } else {
        std::cout << "少し肌寒いかもしれません。" << std::endl;
    }

    // 2. switch文 (多岐分岐)
    // -------------------------------------
    // 特定の変数の値に応じて、処理を分岐させます。
    // `case`の最後には`break`を忘れないようにしましょう。
    int day_of_week = 3; // 1:月曜, 2:火曜, ...

    std::cout << "\n--- switch文の例 --- " << std::endl;
    switch (day_of_week) {
        case 1:
            std::cout << "月曜日です。" << std::endl;
            break;
        case 2:
            std::cout << "火曜日です。" << std::endl;
            break;
        case 3:
            std::cout << "水曜日です。" << std::endl;
            break;
        case 4:
            std::cout << "木曜日です。" << std::endl;
            break;
        case 5:
            std::cout << "金曜日です。" << std::endl;
            break;
        case 6:
        case 7:
            std::cout << "週末です！" << std::endl;
            break;
        default:
            std::cout << "無効な曜日です。" << std::endl;
            break;
    }

    // 3. forループ (繰り返し)
    // -------------------------------------
    // 指定した回数だけ処理を繰り返します。
    std::cout << "\n--- forループの例 --- " << std::endl;
    for (int i = 0; i < 5; ++i) {
        std::cout << "現在のカウント: " << i << std::endl;
    }

    // 4. 範囲ベースのforループ (Range-based for loop)
    // -------------------------------------
    // 配列やベクターなどのコンテナの全要素に対して処理を行います。
    // こちらの方が簡潔で安全です。
    std::vector<std::string> fruits = {"Apple", "Banana", "Cherry"};
    std::cout << "\n--- 範囲ベースforループの例 --- " << std::endl;
    for (const std::string& fruit : fruits) {
        std::cout << "果物: " << fruit << std::endl;
    }

    // 5. whileループ (条件付き繰り返し)
    // -------------------------------------
    // 指定した条件がtrueである間、処理を繰り返します。
    int countdown = 3;
    std::cout << "\n--- whileループの例 --- " << std::endl;
    while (countdown > 0) {
        std::cout << countdown << "..." << std::endl;
        countdown--; // カウントを減らす
    }
    std::cout << "発射！" << std::endl;

    // 6. do-whileループ
    // -------------------------------------
    // 最低1回は実行されるwhileループです。条件判定がループの後に行われます。
    int number;
    std::cout << "\n--- do-whileループの例 --- " << std::endl;
    do {
        std::cout << "正の数を入力してください: ";
        std::cin >> number;
    } while (number <= 0);
    std::cout << "入力された数: " << number << std::endl;

    return 0;
}

/*
練習問題:
1. 1から10までの整数のうち、偶数だけを表示するforループを書いてみましょう。
   ヒント: `%` (剰余) 演算子を使います (例: `i % 2 == 0`)
2. ユーザーに好きな果物の名前をいくつか入力してもらい、"end"と入力されたらループを終了するプログラムをwhileループを使って書いてみましょう。
3. 1から100までの合計を計算して表示するプログラムを書いてみましょう。
*/

File: learning/rust/02_control_flow.rs
fn main() {
    // ==========================================
    // 制御構文
    // ==========================================
    
    println!("=== Rustの制御構文 ===");
    
    // ------------------------------------------
    // if文
    // ------------------------------------------
    
    println!("\n=== if文 ===");
    
    let score = 85;
    
    // 基本的なif文
    if score >= 90 {
        println!("優秀です！");
    } else if score >= 70 {
        println!("良い成績です。");
    } else if score >= 50 {
        println!("合格です。");
    } else {
        println!("がんばりましょう。");
    }
    
    // if式（値を返す）
    let grade = if score >= 90 {
        "A"
    } else if score >= 80 {
        "B"
    } else if score >= 70 {
        "C"
    } else {
        "D"
    };
    println!("成績: {}", grade);
    
    // 複雑な条件
    let temperature = 25;
    let is_sunny = true;
    
    if temperature > 20 && is_sunny {
        println!("散歩にいい天気です！");
    } else if temperature > 20 || is_sunny {
        println!("まあまあの天気です。");
    } else {
        println!("家にいましょう。");
    }
    
    // ------------------------------------------
    // match文
    // ------------------------------------------
    
    println!("\n=== match文 ===");
    
    let day = 3;
    
    // 基本的なmatch文
    match day {
        1 => println!("月曜日"),
        2 => println!("火曜日"),
        3 => println!("水曜日"),
        4 => println!("木曜日"),
        5 => println!("金曜日"),
        6 | 7 => println!("週末"), // 複数の値をまとめて処理
        _ => println!("無効な日付"), // その他すべて
    }
    
    // match式（値を返す）
    let day_type = match day {
        1..=5 => "平日",     // 範囲指定
        6 | 7 => "週末",
        _ => "無効",
    };
    println!("日付タイプ: {}", day_type);
    
    // Option型のmatch
    let maybe_number: Option<i32> = Some(42);
    
    match maybe_number {
        Some(value) => println!("値があります: {}", value),
        None => println!("値がありません"),
    }
    
    // Result型のmatch
    let division_result: Result<i32, String> = divide_safe(10, 2);
    
    match division_result {
        Ok(value) => println!("除算結果: {}", value),
        Err(error) => println!("エラー: {}", error),
    }
    
    // 複雑なパターンマッチング
    let coordinates = (0, -2);
    
    match coordinates {
        (0, 0) => println!("原点です"),
        (0, y) => println!("Y軸上の点: y = {}", y),
        (x, 0) => println!("X軸上の点: x = {}", x),
        (x, y) if x == y => println!("対角線上の点: ({}, {})", x, y),
        (x, y) => println!("一般的な点: ({}, {})", x, y),
    }
    
    // ------------------------------------------
    // loop文
    // ------------------------------------------
    
    println!("\n=== loop文 ===");
    
    // 無限ループ（breakで抜ける）
    let mut counter = 0;
    let result = loop {
        counter += 1;
        
        if counter == 3 {
            println!("カウンター: {}", counter);
            continue; // 次の反復へ
        }
        
        if counter == 5 {
            break counter * 2; // 値を返してループを抜ける
        }
        
        println!("カウンター: {}", counter);
    };
    
    println!("ループの結果: {}", result);
    
    // ネストしたループとラベル
    'outer: loop {
        println!("外側のループに入りました");
        
        loop {
            println!("内側のループに入りました");
            break 'outer; // 外側のループを抜ける
        }
    }
    println!("外側のループを抜けました");
    
    // ------------------------------------------
    // while文
    // ------------------------------------------
    
    println!("\n=== while文 ===");
    
    // 基本的なwhile文
    let mut number = 3;
    
    while number != 0 {
        println!("カウントダウン: {}", number);
        number -= 1;
    }
    println!("発射！");
    
    // 条件付きwhile
    let mut stack = vec![1, 2, 3, 4, 5];
    
    while let Some(top) = stack.pop() {
        println!("スタックからポップ: {}", top);
    }
    
    // ------------------------------------------
    // for文
    // ------------------------------------------
    
    println!("\n=== for文 ===");
    
    // 配列に対するfor文
    let fruits = ["apple", "banana", "cherry", "date"];
    
    println!("果物リスト:");
    for fruit in fruits.iter() {
        println!("- {}", fruit);
    }
    
    // インデックス付きfor文
    println!("インデックス付き:");
    for (index, fruit) in fruits.iter().enumerate() {
        println!("{}: {}", index, fruit);
    }
    
    // 範囲for文
    println!("1から5まで:");
    for number in 1..=5 {
        println!("{}", number);
    }
    
    // 逆順
    println!("5から1まで:");
    for number in (1..=5).rev() {
        println!("{}", number);
    }
    
    // ステップ付き
    println!("2ずつ増加:");
    for number in (0..10).step_by(2) {
        println!("{}", number);
    }
    
    // Vecに対するfor文
    let numbers = vec![10, 20, 30, 40, 50];
    
    println!("ベクターの要素:");
    for number in &numbers {
        println!("{}", number);
    }
    
    // 可変参照でfor文
    let mut mutable_numbers = vec![1, 2, 3, 4, 5];
    
    println!("変更前: {:?}", mutable_numbers);
    for number in &mut mutable_numbers {
        *number *= 2;
    }
    println!("変更後: {:?}", mutable_numbers);
    
    // 所有権を移動するfor文
    let owned_numbers = vec![100, 200, 300];
    
    println!("所有権を移動:");
    for number in owned_numbers {
        println!("{}", number);
    }
    // owned_numbers はここではもう使用できない
    
    // ------------------------------------------
    // if let と while let
    // ------------------------------------------
    
    println!("\n=== if let と while let ===");
    
    // if let - Option型の簡潔な処理
    let some_value = Some(3);
    
    if let Some(x) = some_value {
        println!("値を取得しました: {}", x);
    } else {
        println!("値がありませんでした");
    }
    
    // while let - 連続的な処理
    let mut optional_numbers = vec![Some(1), Some(2), Some(3), None, Some(4)];
    
    while let Some(maybe_number) = optional_numbers.pop() {
        match maybe_number {
            Some(number) => println!("数値: {}", number),
            None => {
                println!("Noneが見つかりました");
                break;
            }
        }
    }
    
    // ------------------------------------------
    // 複雑な制御フロー
    // ------------------------------------------
    
    println!("\n=== 複雑な制御フロー ===");
    
    // ネストしたmatch
    let result: Result<Option<i32>, String> = Ok(Some(42));
    
    match result {
        Ok(Some(value)) => println!("成功して値があります: {}", value),
        Ok(None) => println!("成功しましたが値がありません"),
        Err(error) => println!("エラーが発生しました: {}", error),
    }
    
    // パターンガード付きmatch
    let pair = (2, -2);
    
    match pair {
        (x, y) if x == -y => println!("対称的な値です: {} と {}", x, y),
        (x, y) if x > y => println!("最初の値が大きいです: {} > {}", x, y),
        (x, y) if x < y => println!("二番目の値が大きいです: {} < {}", x, y),
        (x, y) => println!("等しい値です: {} = {}", x, y),
    }
    
    // 構造体のパターンマッチング
    #[derive(Debug)]
    struct Point {
        x: i32,
        y: i32,
    }
    
    let point = Point { x: 0, y: 7 };
    
    match point {
        Point { x, y: 0 } => println!("X軸上の点: x = {}", x),
        Point { x: 0, y } => println!("Y軸上の点: y = {}", y),
        Point { x, y } => println!("その他の点: ({}, {})", x, y),
    }
    
    // 値の束縛
    let message = Some(String::from("Hello"));
    
    match message {
        Some(ref text) if text.len() > 3 => {
            println!("長いメッセージ: {}", text);
        }
        Some(text) => {
            println!("短いメッセージ: {}", text);
        }
        None => println!("メッセージがありません"),
    }
}

fn divide_safe(a: i32, b: i32) -> Result<i32, String> {
    if b == 0 {
        Err(String::from("ゼロで割ることはできません"))
    } else {
        Ok(a / b)
    }
}